avalon/tools/loader/model.py

```diff
@@ -141,17 +150,72 @@ def set_version(self, index, version):
             "step": version_data.get("step", None)
         })
 
-    def refresh(self):
+    def fetch_subset_and_version(self):
+        """Query all subsets and latest versions from aggregation
+
+        (NOTE) The returned version documents are NOT the real version
+            document, it's generated from the MongoDB's aggregation so
+            some of the first level field may not be presented.
+
+        """
+        def _fetch():
+            _subsets = list()
+            _ids = list()
+            for subset in io.find({"type": "subset",
+                                   "parent": self._asset_id}):
+                if self._doc_fetching_stop:
+                    return
+                _subsets.append(subset)
+                _ids.append(subset["_id"])
+
+            _pipeline = [
+                # Find all versions of those subsets
+                {"$match": {"type": "version", "parent": {"$in": _ids}}},
+                # Sorting versions all together
+                {"$sort": {"name": 1}},
+                # Group them by "parent", but only take the last
+                {"$group": {"_id": "$parent",
+                            "_version_id": {"$last": "$_id"},
+                            "name": {"$last": "$name"},
+                            "data": {"$last": "$data"},
+                            "locations": {"$last": "$locations"},
+                            "schema": {"$last": "$schema"}}},
+            ]
+            versions = dict()
+            for doc in io.aggregate(_pipeline):
+                if self._doc_fetching_stop:
+                    return
+                doc["parent"] = doc["_id"]
+                doc["_id"] = doc.pop("_version_id")
+                versions[doc["parent"]] = doc
+
+            self._doc_payload[:] = [(subset, versions.get(subset["_id"]))
+                                    for subset in _subsets]
+            self.doc_fetched.emit()
+
+        self._doc_payload[:] = []
```

Comment:
I am curious why this is used `self._doc_payload[:] = []` and not just `self._doc_payload = []`?

Reply:
This is basically an in-place assignment. It would be the same as doing:
```python
my_list = ["a", "b", "c"]
my_list.clear()
```

I believe the reason to do explicit in-place assignments is due to the "mutable list" potentially being shared to another thread and otherwise that one working on an older copy of a list that isn't actually the same list as `self._doc_payload`. However, because it doesn't seem to be passed around as an object I don't think that's actually the case. @davidlatwe is it redundant for this case?

There should be some informational topics online on replacing lists in-place with regards to its mutability.

Reply:
> However, because it doesn't seem to be passed around as an object I don't think that's actually the case.

Yes, it's true that there's no need to use in-place assignment, just re-using the same object for no special reason. Will change it for good.